aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,502 @@
+			log.Printf("[ERROR] An error occured creating new AWS KMS Grant: %s", err)
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Created new KMS Grant: %s", *out.GrantId)
+	d.SetId(*out.GrantId)
+	d.Set("grant_id", out.GrantId)
+	d.Set("grant_token", out.GrantToken)
+
+	return resourceAwsKmsGrantRead(d, meta)
+}
+
+func resourceAwsKmsGrantRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Id()
+	keyId := d.Get("key_id").(string)
+
+	log.Printf("[DEBUG] Looking for grant id: %s", grantId)
+	grant, err := findKmsGrantByIdWithRetry(conn, keyId, grantId)
+
+	if err != nil {
+		return err
+	}
+
+	if grant == nil {
+		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
+		d.SetId("")
+	}
+
+	if grant != nil {
+		// The grant sometimes contains principals that identified by their unique id: "AROAJYCVIVUZIMTXXXXX"
+		// instead of "arn:aws:...", in this case don't update the state file
+		if strings.HasPrefix(*grant.GranteePrincipal, "arn:aws") {
+			d.Set("grantee_principal", grant.GranteePrincipal)
+		} else {
+			log.Printf(
+				"[WARN] Unable to update grantee principal state %s for grant id %s for key id %s.",
+				*grant.GranteePrincipal, grantId, keyId)
+		}
+
+		if grant.RetiringPrincipal != nil {
+			if strings.HasPrefix(*grant.RetiringPrincipal, "arn:aws") {
+				d.Set("retiring_principal", grant.RetiringPrincipal)
+			} else {
+				log.Printf(
+					"[WARN] Unable to update retiring principal state %s for grant id %s for key id %s",
+					*grant.RetiringPrincipal, grantId, keyId)
+			}
+		}
+
+		if err := d.Set("operations", aws.StringValueSlice(grant.Operations)); err != nil {
+			log.Printf("[DEBUG] Error setting operations for grant %s with error %s", grantId, err)
+		}
+		if *grant.Name != "" {
+			d.Set("name", grant.Name)
+		}
+		if grant.Constraints != nil {
+			if err := d.Set("constraints", flattenKmsGrantConstraints(grant.Constraints)); err != nil {
+				log.Printf("[DEBUG] Error setting constraints for grant %s with error %s", grantId, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Retiring grants requires special permissions (i.e. the
+// caller to be root, retiree principal, or grantee principal with retire grant
+// privileges). So just revoke grants.
+func resourceAwsKmsGrantDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Get("grant_id").(string)
+	keyId := d.Get("key_id").(string)
+	input := kms.RevokeGrantInput{
+		GrantId: aws.String(grantId),
+		KeyId:   aws.String(keyId),
+	}
+
+	log.Printf("[DEBUG] Revoking KMS grant: %s", grantId)
+	_, err := conn.RevokeGrant(&input)
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Checking if grant is revoked: %s", grantId)
+	err = waitForKmsGrantToBeRevoked(conn, keyId, grantId)
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
```

Comment:
Nitpick: `d.SetId("")` is no longer required in delete functions

Reply:
Done. Didn't realize that.